### DISCUSSION TITLE
Performance tips for rendering tiled terrain

### DISCUSSION THREAD

**Discussion by smfloery** [pygfx/pygfx#851] [Q&A]
Dear all,
we are using pgfx to render real world terrain represented as triangulated mesh. So far we only used one mesh for the whole terrain. Obviously, at some point with larger terrains this trivial approach does not work anymore.

Hence, we had the idea of using frustum culling: We split our terrain into several tiles and only display those tiles which are currently visible. This should suffice for the moment; While the frustum culling itself works, the problem we are facing now is within the process of adding the visible tiles step by step: 

After frustum culling we have a list of tiles (`visible_tiles`) which we want to load to the scene. But we want to show the user the progress of loading the tiles and hence, update the canvas after each tile is loaded. For our understanding it should suffice to call 
`canvas.request_draw(lambda: renderer.render(scene, camera))`
within each iteration of the for-loop:

```
canvas = WgpuCanvas()
renderer = gfx.WgpuRenderer(canvas)
scene = gfx.Scene()
camera = gfx.PerspectiveCamera(fov=45, depth_range=(1, 100000))

for tile in visible_tiles:          
  # print("Loading mesh...")
  mesh_geom = gfx.geometries.Geometry(indices=faces, 
                                      positions=verts.astype(np.float32))
  mesh_material = gfx.MeshNormalMaterial(side="FRONT")
  mesh = gfx.Mesh(mesh_geom, mesh_material)
  
  scene.add(mesh)
  # print(len(scene.children))
  canvas.request_draw(lambda: renderer.render(scene, camera))
```
with `faces` and `verts` being numpy arrays containing the respective data for each tile. However, no matter how we do it the canvas is only updated after all tiles have been added. Calling `len(scene.children)` within each iteration shows that the tiles are added one by one. 

Would it be a workaround to add all tiles on startup, but hidden. And only change the visibility of each tile instead of adding them to the scene?

Is it generally possible to load the tiles on the fly such that the user can continue looking around with the camera? Hence, in a non-blocking manner? We are using pygfx in combination with QT follwing this example: https://docs.pygfx.org/stable/_gallery/other/integration_qt.html

All the best
Sebastian

**Comment by almarklein** [comment]
I think setting the objects `.visible` flag would be an easier (and possibly more performant) way to do this, unless gpu memory is a bottleneck too.

As for the update, the `.request_draw()` call requests a draw, which will then happen sometime later in the event-loop. The way to implement what you want you be something like:
```py
tiles_to_add  = []

def animate():
    if tiles_to_add:
        tile = tiles_to_add.pop(0)
        mesh = ....
        scene.add(mesh)
    rendere.render(...)
```

BTW: we have planned changes to update mechanics like this, and also allow direct draws in the future.

**Comment by smfloery** [comment]
Dear almarklen,
thanks for your response and sorry for my delayed answer. I ended up doing something similar to your suggestion which works in general:

![monique_animation_short](https://github.com/user-attachments/assets/c48f07bd-72ed-455a-a18f-7f1c8e685506)

In this animation I use 25 tiles spanning an area of 25x25km. Hence, each tile is 5kmx5km. I add all of these on startup and in the loop, as suggested by you, I only set "visible=True" or "visible=False". Additionally, I also add an Orthophoto as texture on top. For this I use a LOD structure: For each tile I prepared an orthophoto in various resolutions. Depending on the distance from the camera I than select either a higher resolution or lower resolution texture. 

However, user experience is laggy which can also be seen from the FPS. Now I wonder, how to improve the performance! As I don't have much experience with this kind of programming, I'm unsure what I can expect and where there might be room to improve. From your experience: Should it be possible to improve performance or do you think thats already the maximum what to achieve with this kind of approach? Might it be better to reduce the number of tiles (by using a quadtree approach) and/or use a LOD structure for the tiles as well (to reduce the number of triangles in the meshes). Simple "profiling" of my animate function shows, that most time is used in the render.render() call.

All the best
Sebastian  

P.S.: As this is not really an issue anymore, I'm not sure if we should move this conversation to another channel? What do you suggest?

**Reply by almarklein** [reply]
> As this is not really an issue anymore, I'm not sure if we should move this conversation to another channel? What do you suggest?

Good point, I converted it to a discussion.

**Comment by almarklein** [comment]
Since you found that the bottleneck seems to be the rendering, an improvement might be to render less tiles, or indeed have versions of the tiles that contain less vertices. I imagine it should not be too hard to do some experiments to see how these parameters affect performance. 

Note that `renderer.render()` also performs tasks like updating the internal wgpu objects and uploading data to the GPU. I recommend adding this somewhere:
```py
gfx.logger.setLevel("INFO")
```
Which will produce log messages when e.g. pipeline objects are rebuild, or shaders need recompiling. These operations are way more expensive than syncing data.

As for the texture, you can do `gfx.Texture(..., generate_mipmaps=True)` and have pygfx generate a pyramid for you. I've not used that myself, but from what I understand, the shader will automatically sample from the best level. @panxinmiao might know more about how that works (e.g. https://www.w3.org/TR/WGSL/#derivatives, in the part that mentions `textureSample`).

**Reply by panxinmiao** [reply]
> As for the texture, you can do gfx.Texture(..., generate_mipmaps=True) and have pygfx generate a pyramid for you. I've not used that myself, but from what I understand, the shader will automatically sample from the best level.

Yes, when a texture has mipmaps, the `textureSample` function automatically selects the appropriate mipmap level for sampling. If you'd like to control the LOD manually, you can use the `textureSampleLevel` function in the shader (without needing to switch or reload the texture).

There are many potential causes for performance issues, and a detailed analysis is necessary. I believe @almarklein's assumption is quite reasonable—frequent unnecessary rebuilding of GPU resources can significantly impact performance.

Additionally, I’m very interested in your example. If it’s convenient, you can share your code here, and I’ll try to see if I can identify any areas for performance improvement, 😄

**Comment by smfloery** [comment]
Dear @almarklein and @panxinmiao,
thanks both of you for your feedback. I just switched my LOD texture approach to using mipmaps and this already improves the rendering by a few FPS. This is great! If I go further and introduce reduced version of my mesh for tiles further away from the camera: Would you rather create the same mesh geometry multiple times or just use on mesh object for each tile and replace the underlying geometry of data on the fly?

As the rendering is part of larger software I'm writing, I will create a standalone example and upload it in the upcoming days!

**Reply by almarklein** [reply]
> Would you rather create the same mesh geometry multiple times or just use on mesh object for each tile and replace the underlying geometry of data on the fly?

I'm not sure what would be better. Swapping out geometry should also be relatively fast. So don't expect there to be a significant difference in performance. But I'd try it to be sure :)
